process/track/blockNotarizer.go

```diff
@@ -0,0 +1,228 @@
+}
+
+func (bn *blockNotarizer) cleanupNotarizedHeadersBehindNonce(shardID uint32, nonce uint64) {
+	if nonce == 0 {
+		return
+	}
+
+	bn.mutNotarizedHeaders.Lock()
+	defer bn.mutNotarizedHeaders.Unlock()
+
+	notarizedHeaders, ok := bn.notarizedHeaders[shardID]
+	if !ok {
+		return
+	}
+
+	headersInfo := make([]*headerInfo, 0)
+	for _, headerInfo := range notarizedHeaders {
+		if headerInfo.header.GetNonce() < nonce {
+			continue
+		}
+
+		headersInfo = append(headersInfo, headerInfo)
+	}
+
+	bn.notarizedHeaders[shardID] = headersInfo
+}
+
+func (bn *blockNotarizer) displayNotarizedHeaders(shardID uint32, message string) {
+	bn.mutNotarizedHeaders.RLock()
+	defer bn.mutNotarizedHeaders.RUnlock()
+
+	notarizedHeaders, ok := bn.notarizedHeaders[shardID]
+	if !ok {
+		return
+	}
+
+	if len(notarizedHeaders) > 1 {
+		sort.Slice(notarizedHeaders, func(i, j int) bool {
+			return notarizedHeaders[i].header.GetNonce() < notarizedHeaders[j].header.GetNonce()
+		})
+	}
+
+	shouldNotDisplay := len(notarizedHeaders) == 0 ||
+		len(notarizedHeaders) == 1 && notarizedHeaders[0].header.GetNonce() == 0
+	if shouldNotDisplay {
+		return
+	}
+
+	log.Debug(message,
+		"shard", shardID,
+		"nb", len(notarizedHeaders))
+
+	for _, headerInfo := range notarizedHeaders {
+		log.Trace("notarized header info",
+			"round", headerInfo.header.GetRound(),
+			"nonce", headerInfo.header.GetNonce(),
+			"hash", headerInfo.hash)
+	}
+}
+
+func (bn *blockNotarizer) getLastNotarizedHeader(shardID uint32) (data.HeaderHandler, []byte, error) {
+	bn.mutNotarizedHeaders.RLock()
+	defer bn.mutNotarizedHeaders.RUnlock()
+
+	if bn.notarizedHeaders == nil {
+		return nil, nil, process.ErrNotarizedHeadersSliceIsNil
+	}
+
+	headerInfo := bn.lastNotarizedHeaderInfo(shardID)
+	if headerInfo == nil {
+		return nil, nil, process.ErrNotarizedHeadersSliceForShardIsNil
+	}
+
+	return headerInfo.header, headerInfo.hash, nil
+}
+
+func (bn *blockNotarizer) getLastNotarizedHeaderNonce(shardID uint32) uint64 {
+	bn.mutNotarizedHeaders.RLock()
+	defer bn.mutNotarizedHeaders.RUnlock()
+
+	if bn.notarizedHeaders == nil {
+		return 0
+	}
+
+	headerInfo := bn.lastNotarizedHeaderInfo(shardID)
+	if headerInfo == nil {
+		return 0
+	}
+
+	return headerInfo.header.GetNonce()
+}
+
+func (bn *blockNotarizer) lastNotarizedHeaderInfo(shardID uint32) *headerInfo {
+	notarizedHeadersCount := len(bn.notarizedHeaders[shardID])
+	if notarizedHeadersCount > 0 {
+		return bn.notarizedHeaders[shardID][notarizedHeadersCount-1]
+	}
+
+	return nil
+}
+
+func (bn *blockNotarizer) getNotarizedHeader(shardID uint32, offset uint64) (data.HeaderHandler, []byte, error) {
+	bn.mutNotarizedHeaders.RLock()
+	defer bn.mutNotarizedHeaders.RUnlock()
+
+	if bn.notarizedHeaders == nil {
+		return nil, nil, process.ErrNotarizedHeadersSliceIsNil
+	}
+
+	headersInfo := bn.notarizedHeaders[shardID]
+	if headersInfo == nil {
+		return nil, nil, process.ErrNotarizedHeadersSliceForShardIsNil
+	}
+
+	headerInfo := headersInfo[0]
```

Comment:
this code will panic if, for some reason (bug, for example), there is an empty slice on `bn.notarizedHeaders[shardID]` (maybe replace the condition on L171 with `if len(headersInfo) == 0 {`
Also, the function returns the first notarized header id offset is out of bound (returns something else as the name suggests)

Reply:
Refactored